File: skills/auto-dupicate-fixer/scripts/detect.py
#!/usr/bin/env python3

"""
Phase 1: Detect Duplicates (Python)
Scans Python project for duplicate modules, functions, structural similarities
"""

import json
import sys
import os
import hashlib
import ast
from pathlib import Path
from collections import defaultdict

def get_python_files(project_root):
    """Get all .py files excluding __pycache__ and venv"""
    files = []
    for root, dirs, filenames in os.walk(project_root):
        dirs[:] = [d for d in dirs if d not in ('__pycache__', 'venv', '.venv', '.git', 'node_modules')]
        for f in filenames:
            if f.endswith('.py'):
                files.append(os.path.join(root, f))
    return files

def get_file_hash(filepath):
    """MD5 hash of file content"""
    with open(filepath, 'rb') as f:
        return hashlib.md5(f.read()).hexdigest()

def parse_python_ast(filepath):
    """Parse Python file and extract structure"""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            return ast.parse(f.read())
    except SyntaxError:
        return None

def extract_functions_and_classes(tree):
    """Extract function and class names"""
    if not tree:
        return []
    
    items = []
    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.ClassDef)):
            items.append(node.name)
    return items

def detect_python_duplicates(project_root, min_tokens=50):
    """Main detection logic"""
    files = get_python_files(project_root)
    
    results = {
        'timestamp': __import__('datetime').datetime.now().isoformat(),
        'python_files': files,
        'identical': [],
        'structural': [],
        'function_duplicates': []
    }
    
    # Identical file detection
    file_hashes = defaultdict(list)
    for filepath in files:
        try:
            file_hash = get_file_hash(filepath)
            file_hashes[file_hash].append(filepath)
        except Exception as e:
            print(f"Error hashing {filepath}: {e}", file=sys.stderr)
    
    for file_hash, file_list in file_hashes.items():
        if len(file_list) > 1:
            results['identical'].append({
                'hash': file_hash,
                'files': file_list,
                'size': os.path.getsize(file_list[0])
            })
    
    # Structural similarity
    structures = defaultdict(list)
    for filepath in files:
        tree = parse_python_ast(filepath)
        if tree:
            items = extract_functions_and_classes(tree)
            struct_key = tuple(sorted(items))
            structures[struct_key].append(filepath)
    
    for struct_key, file_list in structures.items():
        if len(file_list) > 1 and len(struct_key) >= min_tokens // 10:
            results['structural'].append({
                'functions_classes': list(struct_key),
                'files': file_list
            })
    
    return results

if __name__ == '__main__':
    project_root = sys.argv[1] if len(sys.argv) > 1 else os.getcwd()
    min_tokens = int(sys.argv[2]) if len(sys.argv) > 2 else 50
    
    duplicates = detect_python_duplicates(project_root, min_tokens)
    print(json.dumps(duplicates, indent=2))

File: skills/auto-dupicate-fixer/scripts/detect.js
#!/usr/bin/env node

/**
 * Phase 1: Detect Duplicates
 * Scans project for duplicate files, blocks, and structural similarities
 * Outputs: duplicates.json
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { execSync } = require('child_process');

const projectRoot = process.argv[2] || process.cwd();
const format = process.argv[3] === '--format=json' ? 'json' : 'text';
const minTokens = parseInt(process.argv[4]?.split('=')[1] || '50');

const results = {
  timestamp: new Date().toISOString(),
  projectRoot,
  duplicates: [],
  structural: [],
  identical: [],
  metrics: {}
};

// Scan TS/JS files
function scanTypeScriptJavaScript() {
  const files = execSync(`find ${projectRoot} -type f \\( -name "*.ts" -o -name "*.tsx" -o -name "*.js" -o -name "*.jsx" \\) ! -path "*/node_modules/*" ! -path "*/.git/*"`, { encoding: 'utf-8' }).trim().split('\n').filter(Boolean);
  
  const fileHashes = {};
  const fileContents = {};

  files.forEach(file => {
    try {
      const content = fs.readFileSync(file, 'utf-8');
      const hash = crypto.createHash('md5').update(content).digest('hex');
      
      fileContents[file] = content;
      
      if (!fileHashes[hash]) {
        fileHashes[hash] = [];
      }
      fileHashes[hash].push(file);
    } catch (e) {
      console.error(`Error reading ${file}:`, e.message);
    }
  });

  // Identical files
  Object.entries(fileHashes).forEach(([hash, files]) => {
    if (files.length > 1) {
      results.identical.push({
        hash,
        files,
        size: fs.statSync(files[0]).size
      });
    }
  });

  // Structural similarity via jscpd
  try {
    const jscpdOutput = execSync(`npx jscpd --reporters=json --output ${projectRoot}/.jscpd-report.json ${projectRoot} --ignore=node_modules,dist,.git`, { encoding: 'utf-8', stdio: 'pipe' }).catch(() => '');
    const report = JSON.parse(fs.readFileSync(`${projectRoot}/.jscpd-report.json`, 'utf-8'));
    
    if (report.duplicates) {
      results.duplicates = report.duplicates.map(dup => ({
        files: dup.locations.map(loc => loc.file),
        lines: dup.lines,
        tokens: dup.tokens,
        percentage: dup.percentage
      })).filter(dup => dup.tokens >= minTokens);
    }
  } catch (e) {
    console.warn('jscpd not available, skipping block analysis');
  }

  results.metrics.tsJsFiles = files.length;
  results.metrics.identicalGroups = Object.keys(fileHashes).filter(h => fileHashes[h].length > 1).length;
}

// Scan Python files
function scanPython() {
  try {
    const pythonResult = execSync(`python3 scripts/detect.py "${projectRoot}" ${minTokens}`, { encoding: 'utf-8' });
    const pythonDuplicates = JSON.parse(pythonResult);
    results.python = pythonDuplicates;
  } catch (e) {
    console.warn('Python detection skipped:', e.message);
  }
}

// Main
console.log(`🔍 Scanning ${projectRoot} for duplicates...`);
scanTypeScriptJavaScript();
scanPython();

const outputFile = path.join(projectRoot, '.duplicate-report.json');
fs.writeFileSync(outputFile, JSON.stringify(results, null, 2));

if (format === 'json') {
  console.log(JSON.stringify(results, null, 2));
} else {
  console.log(`\n📊 Detection Summary:`);
  console.log(`  TS/JS Files: ${results.metrics.tsJsFiles}`);
  console.log(`  Identical Groups: ${results.metrics.identicalGroups}`);
  console.log(`  Duplicate Blocks: ${results.duplicates.length}`);
  console.log(`  Report: ${outputFile}`);
}

process.exit(0);
